File: framework/src/main/java/com/github/mrmitew/bodylog/framework/common/view/InjectableViewModel.java
package com.github.mrmitew.bodylog.framework.common.view;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;

import com.github.mrmitew.bodylog.framework.AndroidApplication;
import com.github.mrmitew.bodylog.framework.di.presenter.PresenterHolderInjector;


public abstract class InjectableViewModel extends AndroidViewModel {
    public InjectableViewModel(final Application application) {
        super(application);
        setupActivityComponent();
    }

    private void setupActivityComponent() {
        injectMembers(((AndroidApplication)getApplication()).getPresenterHolderInjector());
    }

    protected abstract void injectMembers(PresenterHolderInjector presenterHolderInjector);
}

File: framework/src/main/java/com/github/mrmitew/bodylog/framework/common/viewmodel/InjectableViewModel.kt
package com.github.mrmitew.bodylog.framework.common.viewmodel

import android.app.Application
import android.arch.lifecycle.AndroidViewModel

import com.github.mrmitew.bodylog.framework.AndroidApplication
import com.github.mrmitew.bodylog.framework.di.presenter.PresenterHolderInjector


abstract class InjectableViewModel(application: Application) : AndroidViewModel(application) {
    init {
        setupActivityComponent()
    }

    private fun setupActivityComponent() =
        injectMembers(AndroidApplication.getPresenterHolderInjector(getApplication()))

    protected abstract fun injectMembers(injector: PresenterHolderInjector)
}